File: backend/src/models/user_model.js
import mongoose from "mongoose";
import bcrypt from "bcrypt";
import jwt from "jsonwebtoken";
import mongooseAggregatePaginate from "mongoose-aggregate-paginate-v2";

const userSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: true,
      trim: true,
    },

    email: {
      type: String,
      required: true,
      unique: true,
      lowercase: true,
    },

    password: {
      type: String,
      required: [true, "Password cannot be empty !!"],
    },

    yearOfAdmission: {
      type: Number,
      required: true,
    },

    isSuperAdmin: {
      type: Boolean,
      default: false,
    },

    isAdmin: {
      type: Boolean,
      default: false,
    },

    isMentorEligible: {
      type: Boolean,
      default: false,
    },

    menteeToMentorId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      default: null,
    },

    mentorToMenteeIds: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: "User",
      },
    ],
    refreshToken: {
      type: String,
    },
  },
  { timestamps: true }
);

userSchema.pre("save", async function (next) {
  if (!this.isModified("password")) return next();
  this.password = await bcrypt.hash(this.password, 10);
  next();
});

userSchema.methods.isPasswordCorrect = async function (password) {
  return await bcrypt.compare(password, this.password);
};

userSchema.methods.generateAccessToken = function () {
  return jwt.sign(
    {
      _id: this.id,
      email: this.email,
    },
    process.env.ACCESS_TOKEN_SECRET,
    {
      expiresIn: process.env.ACCESS_TOKEN_EXPIRY,
    }
  );
};
userSchema.methods.generateRefreshToken = function () {
  return jwt.sign(
    {
      _id: this._id,
    },
    process.env.REFRESH_TOKEN_SECRET,
    {
      expiresIn: process.env.REFRESH_TOKEN_EXPIRY,
    }
  );
};

userSchema.plugin(mongooseAggregatePaginate);
export const User = mongoose.model("User", userSchema);

File: frontend/lib/models/user_model.dart
class User {
  final String id;
  final String name;
  final String email;
  final int yearOfAdmission;
  final bool isSuperAdmin;
  final bool isAdmin;
  final bool isMentorEligible;
  final String? menteeToMentorId;
  final List<String> mentorToMenteeIds;

  User({
    required this.id,
    required this.name,
    required this.email,
    required this.yearOfAdmission,
    required this.isSuperAdmin,
    required this.isAdmin,
    required this.isMentorEligible,
    required this.menteeToMentorId,
    required this.mentorToMenteeIds,
  });

  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      id: json['_id'],
      name: json['name'],
      email: json['email'],
      yearOfAdmission: json['yearOfAdmission'],
      isSuperAdmin: json['isSuperAdmin'],
      isAdmin: json['isAdmin'],
      isMentorEligible: json['isMentorEligible'],
      menteeToMentorId: json['menteeToMentorId'],
      mentorToMenteeIds: List<String>.from(json['mentorToMenteeIds']),
    );
  }
}
